virtio-devices/src/balloon.rs

```diff
@@ -138,6 +193,49 @@ impl BalloonEpollHandler {
     }
 
     fn release_memory_range(
+        &mut self,
+        memory: &GuestMemoryMmap,
+        range_base: GuestAddress,
+        range_len: usize,
+    ) -> result::Result<(), Error> {
+        let page_size: u64 = get_page_size();
+        if page_size == 1 << VIRTIO_BALLOON_PFN_SHIFT {
+            return Self::discard_memory_region(memory, range_base, range_len);
+        }
+
+        if self.pbp.addr == 0 {
+            if range_base.0 & (page_size - 1) != 0 {
+                warn!(
+                    "Can't release memory region: 0x{:x} that doesn't align with page size",
+                    range_base.0
+                );
+                return Ok(());
+            }
+            // Align to page size and set it to pbp.addr
+            self.pbp.addr = align_page_size_up(range_base.0);
```

Comment:
Why do we need to `align_page_size_up()` given the `range_base` is required to be `page_size` aligned (as you did the check few lines above)?

Reply:
Yeah, additional the error you point out I need do some other change here. Maybe we should release memory range one page after another. So,
What about changing it to:

```
        let mut addr = range_base.0 as u64;
        let mut len = range_len as u64;
        while len > 0 {
            if !self.pbp.as_ref().unwrap().pfn_match(addr) {
                self.pbp.as_mut().unwrap().reset();
                // We are trying to free memory region in a different pfn with current pbp. Flush addr
                self.pbp.as_mut().unwrap().addr = align_page_size_up(addr);
                len -= self.pbp.as_ref().unwrap().addr - addr;
                if len <= 0 {
                    return Ok(());
                }
            }

            self.pbp.as_mut().unwrap().set_bit(page_size, addr);
            len -= 1 << VIRTIO_BALLOON_PFN_SHIFT;
            addr += 1 << VIRTIO_BALLOON_PFN_SHIFT;
            if !self.pbp.as_ref().unwrap().bitmap_full() {
                continue;
            }

            Self::discard_memory_region(
                memory,
                vm_memory::GuestAddress(self.pbp.as_ref().unwrap().addr),
                page_size as usize,
            )?;
            self.pbp.as_mut().unwrap().reset();
        }
```

Reply:
It is kind of hard to read in this way without context. Please submit the pages with other updates to this PR. We can discuss from there. Thank you.

Reply:
OK.